play-services-core/src/main/kotlin/com/google/android/gms/locationsharingreporter/service/LocationSharingReporterExtensions.kt

```diff
@@ -0,0 +1,309 @@
+        } else {
+            store
+        }
+    }
+}
+
+fun requestUploadLocation(
+        context: Context,
+        account: Account,
+        request: UploadLocationRequest
+): UploadLocationResponse? {
+    return runCatching {
+        val grpcClient: UserLocationFrontendServiceClient = grpcClient(account, AccountManager.get(context))
+        grpcClient.UploadLocation().executeBlocking(request)
+    }.onFailure { exception ->
+        Log.e(TAG, "Upload location failed", exception)
+    }.getOrNull()
+}
+
+fun validateGoogleAccount(account: Account) {
+    require(account.type == AuthConstants.DEFAULT_ACCOUNT_TYPE) {
+        "Invalid account type, not a Google account"
+    }
+}
+
+fun validateMakePrimaryOption(option: Int) {
+    val validOptions = listOf(
+            ReportingType.ONGOING_REPORTING_ENABLED.value,
+            ReportingType.SINGLE_SHARE_REPORTING_ENABLED.value,
+    )
+    require(option in validOptions) { "Invalid MakePrimaryOption" }
+}
+
+fun validateReportingType(type: Int) {
+    val validTypes = listOf(
+            ReportingType.SINGLE_SHARE_REPORTING_ENABLED.value,
+            ReportingType.ONGOING_REPORTING_ENABLED.value,
+    )
+    require(type in validTypes) {
+        "Invalid ReportingType, valid reporting types are: SINGLE_SHARE_REPORTING_ENABLED, ONGOING_REPORTING_ENABLED."
+    }
+}
+
+fun validateLocationShare(locationShare: LocationShare?) {
+    require(locationShare?.obfuscatedGaiaId != null || locationShare?.tokenId != null) {
+        "Invalid location share, neither obfuscated gaia ID nor token ID are specified in it."
+    }
+}
+
+fun updateDeviceBatterySaverState(context: Context) {
+    val powerManager = context.getSystemService(Context.POWER_SERVICE) as PowerManager
+    val isPowerSaveMode = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+        powerManager.isPowerSaveMode
+    } else {
+        false
+    }
+    Log.i(TAG, "Updating device battery saver state, isEnabled: $isPowerSaveMode")
+
+    ReportingRequestStoreFile.updateReportingRequestStore(context) { requestStore ->
+        val state = if (isPowerSaveMode) STATE_ENABLED else STATE_DISABLED
+        return@updateReportingRequestStore requestStore.newBuilder().batterySaverState(state).build()
+    }
+}
+
+fun updateDeviceLocationSettingState(context: Context) {
+    val locationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
+    val isEnabled = LocationManagerCompat.isLocationEnabled(locationManager)
+    Log.i(TAG, "Updating device location setting state, isEnabled: $isEnabled")
+
+    ReportingRequestStoreFile.updateReportingRequestStore(context) { requestStore ->
+        val state = if (isEnabled) STATE_ENABLED else STATE_DISABLED
+        return@updateReportingRequestStore requestStore.newBuilder().locationSettingState(state).build()
+    }
+}
+
+object ReportingObject {
+    val generalIssues = HashSet<Int>()
+    val issuesByAccount = HashMap<String, HashSet<Int>>()
+}
```

Comment:
Don't use singleton objects except where really applicable.

This one should be a class returned from `getLocationReportingStatus` instead, otherwise `ReportingObject.generalIssues` will stuck with previous issues until device restart even if they are no longer valid.